packages/oauth/src/install-provider.ts

```diff
@@ -336,48 +436,65 @@ export class InstallProvider {
-        const searchParams = InstallProvider.extractSearchParams(req);
+        const searchParams = extractSearchParams(req);
         flowError = searchParams.get('error') as string;
         if (flowError === 'access_denied') {
           throw new AuthorizationError('User cancelled the OAuth installation flow!');
         }
         code = searchParams.get('code') as string;
-        state = searchParams.get('state') as string;
+        stateInQueryString = searchParams.get('state') as string;
         if (!code) {
           throw new MissingCodeError('Redirect url is missing the required code query parameter');
         }
-        if (this.stateVerification && !state) {
+        if (this.stateVerification && !stateInQueryString) {
           throw new MissingStateError('Redirect url is missing the state query parameter. If this is intentional, see options for disabling default state verification.');
         }
       } else {
         throw new UnknownError('Something went wrong');
       }
       // If state verification is enabled, attempt to verify, otherwise ignore
-      if (this.stateVerification && this.stateStore) {
-        // eslint-disable-next-line no-param-reassign
-        installOptions = await this.stateStore.verifyStateParam(new Date(), state);
+      if (this.stateVerification) {
+        try {
+          if (this.legacyStateVerification) {
+            // This mode is not enabled by default
+            // This option is for some of the existing developers that need time for migration
+            this.logger.warn('Enabling legacyStateVerification is not recommended as it does not properly work for OAuth CSRF protection. Please consider migrating from directly using InstallProvider#generateInstallUrl() to InstallProvider#handleInstallPath() for serving the install path.');
+          } else {
+            const stateInBrowserSession: string | undefined = extractCookieValue(req, this.stateCookieName);
+            if (!stateInBrowserSession || (stateInBrowserSession !== stateInQueryString)) {
+              throw new InvalidStateError('The state parameter is not for this browser session.');
+            }
+          }
+          if (this.stateStore) {
+            // eslint-disable-next-line no-param-reassign
+            installOptions = await this.stateStore.verifyStateParam(new Date(), stateInQueryString);
+          } else {
+            throw new InstallerInitializationError('StateStore is not properly configured');
+          }
+        } finally {
+          // Delete the state value in cookies in any case
+          res.setHeader('Set-Cookie', this.buildSetCookieHeaderForStateDeletion());
+        }
       }
       if (!installOptions) {
         const emptyInstallOptions: InstallURLOptions = { scopes: [] };
         // eslint-disable-next-line no-param-reassign
         installOptions = emptyInstallOptions;
       }
 
-      const client = new WebClient(undefined, this.clientOptions);
```

Comment:
Removed this instance creation and changed the following code to use a singleton object.